app/validators/value_source_validator.py

```diff
@@ -26,6 +55,118 @@ def __init__(self, value_source, json_path, questionnaire_schema):
             "calculated_summary": self.questionnaire_schema.calculated_summary_block_ids,
         }
 
+        if self.value_source["source"] == "progress":
+            self._valid_source_identifiers_map["progress"] = {
+                "block_ids": self._get_valid_progress_value_source_block_identifiers(),
+                "section_ids": self._get_valid_progress_value_source_section_identifiers(),
+            }
+
+    def _get_valid_progress_value_source_block_identifiers(self):
+        valid_block_ids = (
+            self.past_block_ids
+            - set([self.current_block_id])
+            - self.block_ids_in_past_repeating_sections
+        )
+
+        return valid_block_ids
+
+    def _get_valid_progress_value_source_section_identifiers(self):
+        valid_section_ids = (
+            self.past_section_ids
+            - set([self.current_section_id])
+            - self.past_repeating_section_ids
+        )
+
+        return valid_section_ids
+
+    @cached_property
+    def block_ids_in_past_repeating_sections(self) -> set[str]:
+        """
+        Returns a list of block IDs that are in repeating sections that are before the current parent section
+        """
+        return set(
+            block["id"]
+            for section_id in self.past_repeating_section_ids
+            for block in self.questionnaire_schema.blocks_by_section_id[section_id]
+        )
+
+    @cached_property
+    def past_repeating_section_ids(self) -> set[str]:
+        """
+        Returns a list of repeating sections' IDs that are before the current parent section
+        """
+        return set(
```

Comment:
Can be updated as above without calling set()

Reply:
Changed